### ISSUE TITLE
pending commit data lock process may be error

### ISSUE THREAD

**Issue by RayYe586** [prometheus/prometheus#9982]
https://github.com/prometheus/prometheus/blob/c965a7555b7ffcee1a127d782abd5bb478a16750/tsdb/head_append.go#L452

	for i, s := range a.samples {
		series = a.sampleSeries[i]
		series.Lock()
		ok, chunkCreated := series.append(s.T, s.V, a.appendID, a.head.chunkDiskMapper)
		series.cleanupAppendIDsBelow(a.cleanupAppendIDsBelow)
		series.pendingCommit = false   <=== **pendingCommint should be set after all series is appended.Gc process may be error **
		series.Unlock()

		if !ok {
			total--
			a.head.metrics.outOfOrderSamples.Inc()
		}
		if chunkCreated {
			a.head.metrics.chunks.Inc()
			a.head.metrics.chunksCreated.Inc()
		}
	}

## process below may be error 

func (s *stripeSeries) gc(mint int64) (map[uint64]struct{}, int, int64) {
	var (
		deleted                  = map[uint64]struct{}{}
		deletedForCallback       = []labels.Labels{}
		rmChunks                 = 0
		actualMint         int64 = math.MaxInt64
	)
	// Run through all series and truncate old chunks. Mark those with no
	// chunks left as deleted and store their ID.
	for i := 0; i < s.size; i++ {
		s.locks[i].Lock()

		for hash, all := range s.hashes[i] {
			for _, series := range all {
				series.Lock()
				rmChunks += series.truncateChunksBefore(mint)

				if len(series.mmappedChunks) > 0 || series.headChunk != nil || series.pendingCommit {
					seriesMint := series.minTime()
					if seriesMint < actualMint {
						actualMint = seriesMint
					}
					series.Unlock()
					continue
				}

**Comment by roidelapluie** [comment]
Thanks for this. I will let @codesome comment on this.

**Comment by codesome** [comment]
@RayYe586 could you please describe what kind of error it can cause?

**Comment by RayYe586** [comment]
@codesome If data commit is in progress,gc may delete these data.I find the error by code analysis.

**Comment by codesome** [comment]
That's true, if we have samples incoming for the gc range, then gc can delete it. The way we take care of it is to not append any samples that gc would delete eventually. It is taken care by [this](https://github.com/prometheus/prometheus/blob/e226d1f95c8abeb7138e065bde13d9f70971288a/tsdb/head_append.go#L138-L142).

The `CompactHead()`, if being used as a library, is required to be used only after stopping all the ingestion. I have updated the issue title since this is more of updating the code comments for DB.CompactHead() and Head.gc()

**Comment by beorn7** [comment]
Hello from the bug scrub!

@codesome it seems easy to amend the code comments as you described above. Could you just do that? (We were tempted to do it ourselves right during the bug scrub, but we were afraid of introducing inaccuracies…)

**Comment by aamoghS** [comment]
I'll document the CompactHead library contract (stop ingestion first) and how Head.gc stays safe with concurrent appends via appendableMinValidTime and pending-commit series.
